py: Change MP_UNARY_OP_INT to MP_UNARY_OP_INT_MAYBE.

dpgeorge opened the pull request:
To make it consistent with MP_UNARY_OP_INT_FLOAT and MP_UNARY_OP_INT_COMPLEX, and allow `int()` to first check if a type supports `__int__` before trying other things (as per CPython).

See related #11598.

dpgeorge committed:
py: Change MP_UNARY_OP_INT to MP_UNARY_OP_INT_MAYBE.

To be consistent with MP_UNARY_OP_INT_FLOAT and MP_UNARY_OP_INT_COMPLEX,
and allow int() to first check if a type supports __int__ before trying
other things (as per CPython).

Signed-off-by: Damien George <damien@micropython.org>

dlech commented on py/objtype.c:
```diff
@@ -432,7 +432,7 @@ STATIC mp_obj_t instance_unary_op(mp_unary_op_t op, mp_obj_t self_in) {
+            case MP_UNARY_OP_INT_MAYBE:
                 // Must return int
                 if (!mp_obj_is_int(val)) {
                     mp_raise_TypeError(NULL);
```
Should this be changed to `return MP_OBJ_NULL` now instead of raising to match the semantics of "maybe"?

dpgeorge replied:
If we get here then the user class defined `__int__` so it does support the method.  But it returned a non-integer type which is illegal.  Hence the raising of an error.

dlech commented on py/objint.c:
```diff
@@ -62,8 +62,17 @@ STATIC mp_obj_t mp_obj_int_make_new(const mp_obj_type_t *type_in, size_t n_args,
             } else {
-                return mp_unary_op(MP_UNARY_OP_INT, args[0]);
+                mp_raise_TypeError_int_conversion(args[0]);
+                #if 0
```
Do we need to keep this `#if 0` block?

dpgeorge replied:
oops, I didn't mean to leave this here

dlech reviewed (commented):
Much better than the `uctypes.PTR` hack.

dpgeorge committed:
py/obj: Accept user types in mp_obj_get_int_maybe.

This is possible now that MP_UNARY_OP_INT_MAYBE exists.

As a consequence mp_obj_get_int now also supports user types, which was
previously possible with MP_UNARY_OP_INT but no tests existed for it.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/basics: Remove __index__ and __inv__ from special methods tests.

MicroPython does not support these special methods, and they may get in the
way of other tests (eg indexing with __int__).

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge commented:
I ran the performance benchmark test suite on PYBv1.0 with this PR and there is effectively no change to the performance scores.
